### ISSUE TITLE
Class::Struct objects break UNIVERSAL::isa

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2009]
Migrated from [rt.perl.org#3284](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3284) (status was 'resolved')

Searchable as RT3284$

**Comment by p5pRT** [comment]
### From pharkins@etoys.com

#### Created by perrin@primenet.com


Objects created using Class&#8203;::Struct return true the first time they are
called with $obj\->isa\('UNIVERSAL'\) but false every time after that\.  I
couldn't get this behavior to happen with objects created in any other
way\.  I even tried putting the code generated by Class&#8203;::Struct directly in
my program in lieu of calling Class&#8203;::Struct\, and it did not exhibit this
behavior\.

Here is a test case that demonstrates the problem&#8203;:

\#\!/usr/bin/perl \-w
use strict;
use Class&#8203;::Struct;

struct\( 'Foo' => \{
&nbsp;          bar       => '$'\, \#' help emacs
&nbsp;         \}\);

my $foo =  bless\( \{\}\, 'Foo' \);

for \(1\.\.2\) \{
&nbsp;    print '$foo\->isa\(\\'Foo\\'\)&#8203;: '       \. $foo\->isa\('Foo'\)       \. "\\n";
&nbsp;    print '$foo\->isa\(\\'UNIVERSAL\\'\)&#8203;: ' \. $foo\->isa\('UNIVERSAL'\) \. "\\n";
\}

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Configured by root at Wed Feb  2 15:34:50 EST 2000.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.2.5-22smp, archname=i386-linux
    uname='linux porky.devel.redhat.com 2.2.5-22smp #1 smp wed jun 2 09:11:51 edt 1999 i686 unknown '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='-O2 -m486 -fno-strength-reduce', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    ccflags ='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    stdchar='char', d_stdstdio=undef, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lposix -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00503:
    /home/pharkins/cvs/webapps/lib/perl
    /home/pharkins/cvs/webapps/conf
    /usr/lib/perl5/5.00503/i386-linux
    /usr/lib/perl5/5.00503
    /usr/lib/perl5/site_perl/5.005/i386-linux
    /usr/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/home/pharkins
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin:/home/pharkins/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin
    PERL5LIB=:/home/pharkins/cvs/webapps/lib/perl:/home/pharkins/cvs/webapps/conf
    PERL_BADLANG (unset)
    SHELL=/bin/bash



```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

#### Created by perrin@primenet.com


Objects created using Class&#8203;::Struct return true the first time they are
called with $obj\->isa\('UNIVERSAL'\) but false every time after that\.  I
couldn't get this behavior to happen with objects created in any other
way\.  I even tried putting the code generated by Class&#8203;::Struct directly in
my program in lieu of calling Class&#8203;::Struct\, and it did not exhibit this
behavior\.

Here is a test case that demonstrates the problem&#8203;:

\#\!/usr/bin/perl \-w
use strict;
use Class&#8203;::Struct;

struct\( 'Foo' => \{
&nbsp;          bar       => '$'\, \#' help emacs
&nbsp;         \}\);

my $foo =  bless\( \{\}\, 'Foo' \);

for \(1\.\.2\) \{
&nbsp;    print '$foo\->isa\(\\'Foo\\'\)&#8203;: '       \. $foo\->isa\('Foo'\)       \. "\\n";
&nbsp;    print '$foo\->isa\(\\'UNIVERSAL\\'\)&#8203;: ' \. $foo\->isa\('UNIVERSAL'\) \. "\\n";
\}

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Configured by root at Wed Feb  2 15:34:50 EST 2000.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.2.5-22smp, archname=i386-linux
    uname='linux porky.devel.redhat.com 2.2.5-22smp #1 smp wed jun 2 09:11:51
edt 1999 i686 unknown '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='-O2 -m486 -fno-strength-reduce',
gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    ccflags ='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    stdchar='char', d_stdstdio=undef, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lposix -lcrypt
    libc=, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00503:
    /home/pharkins/cvs/webapps/lib/perl
    /home/pharkins/cvs/webapps/conf
    /usr/lib/perl5/5.00503/i386-linux
    /usr/lib/perl5/5.00503
    /usr/lib/perl5/site_perl/5.005/i386-linux
    /usr/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/home/pharkins
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
   
PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin:/home/pharkins/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin:/usr/local/bin:/opt/netscape:/home/oracle/bin
   
PERL5LIB=:/home/pharkins/cvs/webapps/lib/perl:/home/pharkins/cvs/webapps/conf
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From perl@tmtm.com

Resolved by change \#20168\.

**Comment by p5pRT** [comment]
@cwest - Status changed from 'open' to 'resolved'
